src/world/Manager/MapMgr.cpp

```diff
@@ -0,0 +1,889 @@
+                newMapData.insert( eventData );
+            }
+            else
+            {
+              newMapData.insert( eventData );
+            }
+          }
+          else if( eventData.first == 393217 || eventData.first == 393223 || eventData.first == 393225 ) // Leve npc locations: Bentbranch / Horizon / Swiftperch
+          {
+            if( player.isQuestCompleted( 220 ) || player.isQuestCompleted( 687 ) || player.isQuestCompleted( 693 ) )
+              newMapData.insert( eventData );
+          }
+        }
+
+        break;
+      }
+
+      case EventHandler::EventHandlerType::CustomTalk:
+      {
+        newMapData.insert( eventData );
+
+        break;
+      }
+
+      case EventHandler::EventHandlerType::GuildOrderGuide:
+      {
+        if( player.isActionLearned( 7 ) )
+          newMapData.insert( eventData );
+
+        break;
+      }
+
+      case EventHandler::EventHandlerType::TripleTriad:
+      {
+        if( eventData.first == 2293771 ) // Triple Triad Master npc for now only
+          newMapData.insert( eventData );
+
+        break;
+      }
+    }
+  }
+  
+  sendPackets( player, newMapData, All );
+}
+
+void Sapphire::World::Manager::MapMgr::updateQuests( Entity::Player& player )
+{
+  auto& mapData = m_mapData[ player.getTerritoryTypeId() ];
+  std::multimap< uint32_t, EventData, less > newMapData;
+
+  for( auto& eventData : mapData )
+  {
+    if( ( eventData.first >> 16 ) == static_cast< uint16_t >( EventHandler::EventHandlerType::Quest ) )
+    {
+      if( isQuestAvailable( player, eventData ) )
+        newMapData.insert( eventData );
+    }
+  }
+  
+  sendPackets( player, newMapData, Quest );;
+}
+
+bool Sapphire::World::Manager::MapMgr::isQuestAvailable( Entity::Player& player, std::pair< const uint32_t, EventData >& eventData )
+{
+  auto& exdData = Common::Service< Data::ExdDataGenerated >::ref();
+
+  auto& quest = m_questData[ eventData.first ];
+
+  if( ( player.isQuestCompleted( eventData.first ) && ( !quest.isRepeatable && eventData.first != 67114 ) ) || player.hasQuest( eventData.first ) ||
+      ( quest.repeatIntervalType == 1 && quest.questRepeatFlag == 0 ) ) // Don't show daily beast tribe quests on the map yet.
+    return false;
+
+  if( quest.classJobUnlock )
+  {
+    if( quest.column18 == 3 )
+      if( static_cast< uint8_t >( player.getClass() ) != quest.classJobUnlock )
+        return false;
+    else if( quest.column18 == 4 )
+      if ( static_cast< uint8_t >( player.getClass() ) == quest.classJobUnlock )
+        return false;
+    else
+      return false;
+  }
+
+  // Was this really ever used?
+  if( quest.startTown )
+  {
+    if( quest.startTown != player.getStartTown() )
+      return false;
+  }
+
+  if( Common::CURRENT_EXPANSION_ID < quest.expansion )
+    return false;
+
+  if( quest.mount )
+  {
+    if( !player.hasMount( quest.mount ) )
+      return false;
+  }
+
+  if( eventData.first == 65968 ) // Quest: A Legend for a Legend
```

Comment:
things like this and basically everything after it should be deferred to a script, not inline

Reply:
like an availability check in a quest script?

Reply:
yeah, probably something like `bool showMapIcon(Entity::Player&)`